refactor: Migrate Resource: mongodbatlas_project to Terraform Plugin Framework and remove api_keys attribute

maastha opened the pull request:
## Description

Migrate Resource: mongodbatlas_project to Terraform Plugin Framework.
- This PR also removes api_keys attribute from project resource.

Link to any related issue(s): [INTMDB-908](https://jira.mongodb.org/browse/INTMDB-908)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [X] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments

maastha committed:
project resource migrated initial

maastha committed:
temp

maastha committed:
resource migrate complete

maastha committed:
data source cleanup

maastha committed:
data source cleanup

maastha committed:
migration tests

AgustinBettati committed:
feat: define new framework provider, configure main and acceptance tests to use mux server with existing sdk v2 provider

AgustinBettati committed:
feat: define new framework provider, configure main and acceptance tests to use mux server with existing sdk v2 provider

AgustinBettati committed:
fix ConflictsWith attribute validation and parsing of list values in assume role schema

AgustinBettati committed:
fix ConflictsWith attribute validation and parsing of list values in assume role schema

maastha committed:
data sources

maastha committed:
data sources

maastha committed:
Merge branch 'INTMDB-971' into INTMDB-908-migrate-project

maastha committed:
integrated changes

maastha committed:
fix tests

maastha committed:
fix linting

maastha committed:
minor

maastha committed:
minor

maastha committed:
lint fix

maastha committed:
minor

AgustinBettati commented on mongodbatlas/provider_test.go:
```diff
@@ -30,6 +35,14 @@ func init() {
 		},
 	}
+
+	config := Config{
+		PublicKey:    os.Getenv("MONGODB_ATLAS_PUBLIC_KEY"),
+		PrivateKey:   os.Getenv("MONGODB_ATLAS_PRIVATE_KEY"),
+		BaseURL:      os.Getenv("MONGODB_ATLAS_BASE_URL"),
+		RealmBaseURL: os.Getenv("MONGODB_REALM_BASE_URL"),
+	}
+	testMongoDBClient, _ = config.NewClient(context.Background())
```
I think here we can leave this testMongoDBClient and remove the testAccProviderSdkV2 which is being used for the same purpose

maastha replied:
That is still used in existing un-migrated resource tests right?

AgustinBettati replied:
Correct. Essentially we can leave either `testMongoDBClient`, or the usage of `testAccProviderSdkV2.Meta().(*MongoDBClient)`. Between both I prefer testMongoDBClient.

maastha replied:
I don't think we should change that in all the other resource and data source tests unless required as we'd be running all the tests then. At least not as part of this PR. We can address that separately.

AgustinBettati replied:
Thats okay, in that case I would however remove the created `testMongoDBClient` in favor of the existing `testAccProviderSdkV2.Meta().(*MongoDBClient)` that is being used in acceptance tests, mainly for consistency.

maastha replied:
@AgustinBettati While that would be consistent in the sense that we use only `testAccProviderSdkV2.Meta().(*MongoDBClient)` throughout our acceptance tests to get the client, it would be inconsistent and incorrect in context of a **Framework-based migrated** resource and it's tests to continue using an SDKv2 based entity. This would also require us to update all the tests once we are ready to get rid of everything that SDKv2 in our provider.

We can instead rename `testMongoDBClient` to  something like `testMongoDBClient_Fw` (open to suggestions for the name) to clarify that this is specific to the Framework migrated resources.
Let me know what you think.

AgustinBettati replied:
If we will progressively transition from testAccProviderSdkV2.Meta().(*MongoDBClient) to using the new testMongoDBClient then I think the existing comments are sufficient

maastha commented on mongodbatlas/fw_data_source_mongodbatlas_projects.go:
```diff
@@ -0,0 +1,202 @@
+
+	stateModel.Results = results
+	stateModel.TotalCount = types.Int64Value(int64(projectsRes.TotalCount))
+	stateModel.ID = types.StringValue(fmt.Sprint(projectsRes.TotalCount))
```
This is a dummy value pretty much because acceptance tests require every data source to have an 'id' field. TF will get rid of it at some point but for now they recommend to just add it and include a deprecation note to ensure users don't use it https://github.com/hashicorp/terraform-plugin-testing/issues/84#issuecomment-1480006432

andreaangiolillo replied:
Can we create a ticket to track this work (remove the ID once the issue is solved)? Also, instead of using totalCount can we use `id.UniqueId()`?

maastha replied:
I'm not sure how id.UniqueId() works, but I will give that a try

maastha replied:
updated to use  id.UniqueId()

AgustinBettati commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	if !projectPlan.IsCollectDatabaseSpecificsStatisticsEnabled.IsUnknown() {
+		projectSettings.IsCollectDatabaseSpecificsStatisticsEnabled = projectPlan.IsCollectDatabaseSpecificsStatisticsEnabled.ValueBoolPointer()
+	}
+	if !projectPlan.IsDataExplorerEnabled.IsUnknown() {
+		projectSettings.IsDataExplorerEnabled = projectPlan.IsDataExplorerEnabled.ValueBoolPointer()
+	}
+	if !projectPlan.IsExtendedStorageSizesEnabled.IsUnknown() {
+		projectSettings.IsExtendedStorageSizesEnabled = projectPlan.IsExtendedStorageSizesEnabled.ValueBoolPointer()
+	}
+	if !projectPlan.IsPerformanceAdvisorEnabled.IsUnknown() {
+		projectSettings.IsPerformanceAdvisorEnabled = projectPlan.IsPerformanceAdvisorEnabled.ValueBoolPointer()
+	}
+	if !projectPlan.IsRealtimePerformancePanelEnabled.IsUnknown() {
+		projectSettings.IsRealtimePerformancePanelEnabled = projectPlan.IsRealtimePerformancePanelEnabled.ValueBoolPointer()
+	}
+	if !projectPlan.IsSchemaAdvisorEnabled.IsUnknown() {
+		projectSettings.IsSchemaAdvisorEnabled = projectPlan.IsSchemaAdvisorEnabled.ValueBoolPointer()
+	}
```
For these optional fields, would .IsNull() be the more appropriate option? Using as reference: https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes#null-and-unknown-values

maastha replied:
no, I used .IsNull() initially but that didn't work as expected because we also have planmodifier for Unknown state for these fields (so these are basically considered unknown and not null)

AgustinBettati commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+			return fmt.Errorf("error adding limits into the project during update: %v", err.Error())
+		}
+	}
+	return nil
```
would give a shot at implementing the previous getChangesInSet, to avoid having to remove and create all elements of the set for a single change

maastha replied:
yeah let me give that a try

maastha replied:
implemented getChangesInSet for teams and limits

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+	client *MongoDBClient
+}
+
+type tfProjectDSModel struct {
```
Any reason why this can not be projectModel?
```suggestion
type projectModel struct {
```

maastha replied:
That was to avoid confusion with atlas models, any models/types that might be created only for tests or temporarily otherwise for any computation. I initally did have `projectModel` but overall I found `tf` prefix to be much more readable especially when dealing with conversion functions, etc, that this is specific to the actual terraform plan/state/config model in question.

These models contain attributes that are Terraforms types, so these are not normal Go models.

Also, since right now we have everything in one package we can end up with naming conflicts when we start adding models for all resources as well as their corresponding data sources. So I think we should definitely keep these separate.

AgustinBettati replied:
As mentioned [here](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1366#discussion_r1284562820), I have found the `tf` prefix helpful when reading and developing.

maastha replied:
Discussed this with @andreaangiolillo offline as well. We will continue using `tf` an `fw` naming conventions. 
This will also be captured in this migration wiki as well.

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+	OrgID                                       types.String    `tfsdk:"org_id"`
+	Created                                     types.String    `tfsdk:"created"`
+	ID                                          types.String    `tfsdk:"id"`
+	Limits                                      []tfLimitModel  `tfsdk:"limits"`
```
```suggestion
	Limits                                      []*tfLimitModel  `tfsdk:"limits"`
```

maastha replied:
updated

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+	Created                                     types.String    `tfsdk:"created"`
+	ID                                          types.String    `tfsdk:"id"`
+	Limits                                      []tfLimitModel  `tfsdk:"limits"`
+	Teams                                       []tfDSTeamModel `tfsdk:"teams"`
```
```suggestion
	Teams                                       []*tfDSTeamModel `tfsdk:"teams"`
```

maastha replied:
updated

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+	IsDataExplorerEnabled                       types.Bool      `tfsdk:"is_data_explorer_enabled"`
+}
+
+type tfDSTeamModel struct {
```
```suggestion
type teamModel struct {
```

maastha replied:
Resolving as we will be using this naming convention as discussed.

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+
+func (d *ProjectDS) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_project"
+}
```
why do we need this?

maastha replied:
The Framework uses Metadata() method to define resource name. In this case resp.TypeName will be `mongodbatlas_project` where `mongodbatlas` is the req.ProviderTypeName defined in the provider.

andreaangiolillo replied:
Can we update this to make project a constant and use format? `("%s_%s, req.ProviderTypeName, projectResourceName)`

maastha replied:
updated

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+	resp.TypeName = req.ProviderTypeName + "_project"
+}
+
+func (d *ProjectDS) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
```
Are we forced to call this function Configure? In generate a better go patter would be to call it newClient and return the client to set for the struct. Feel free to skip this suggestion if Configure is part of an interface

maastha replied:
Yes. Configure is part of an interface.

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+		}
+	}
+
+	// GET PROJECT PROPS
```
```suggestion
```

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+		project *matlas.Project
+	)
+
+	if !projectState.ProjectID.IsNull() {
+		projectID := projectState.ProjectID.ValueString()
+		project, _, err = conn.Projects.GetOneProject(ctx, projectID)
+		if err != nil {
+			resp.Diagnostics.AddError("error when getting project from Atlas", fmt.Sprintf(errorProjectRead, projectID, err.Error()))
+			return
+		}
```
why the projectID is null? Shouldn't always be set as part of the create?

maastha replied:
This is the data source, so the user can set either name or project_id, both of which are optional and we query accordingly based on what the user has provided. This logic has been migrated as is from the previous data source.
https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/data-sources/project#argument-reference

andreaangiolillo replied:
Thanks

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+
+	projectState = toTFProjectDataSourceModel(ctx, project, atlasTeams, atlasProjectSettings, atlasLimits)
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &projectState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```
is this a patter of the framework? I see that we don't return error anymore 🤔

maastha replied:
Right. We don't have to return the error anymore, but only add them to `resp.Diagnostics` and return

andreaangiolillo replied:
Thanks for confirming this

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+func toTFProjectDataSourceModel(ctx context.Context, project *matlas.Project,
+	teams *matlas.TeamsAssigned, projectSettings *matlas.ProjectSettings, limits []admin.DataFederationLimit) tfProjectDSModel {
+	projectStateModel := tfProjectDSModel{
+		ID:           types.StringValue(project.ID),
+		ProjectID:    types.StringValue(project.ID),
```
any reason why we have both id and projectID for the same value?

maastha replied:
Yes. // https://github.com/hashicorp/terraform-plugin-testing/issues/84#issuecomment-1480006432
Every data source needs to have an 'id' value. As per comment here https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1365#discussion_r1284372784

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+func toTFProjectDataSourceModel(ctx context.Context, project *matlas.Project,
+	teams *matlas.TeamsAssigned, projectSettings *matlas.ProjectSettings, limits []admin.DataFederationLimit) tfProjectDSModel {
```
Can the projectSettings be null? if yes, we are going to get null pointers errors in line 218

maastha replied:
projectSettings is always retrieved from `getProjectPropsFromAPI()` where we do check if the API ever returns an error so this cannot be null.

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,261 @@
+		Limits:                                      toTFLimitsDataSourceModel(ctx, limits),
+	}
+
+	return projectStateModel
```
```suggestion
	return tfProjectDSModel{
		ID:           types.StringValue(project.ID),
		ProjectID:    types.StringValue(project.ID),
		Name:         types.StringValue(project.Name),
		OrgID:        types.StringValue(project.OrgID),
		ClusterCount: types.Int64Value(int64(project.ClusterCount)),
		Created:      types.StringValue(project.Created),
		IsCollectDatabaseSpecificsStatisticsEnabled: types.BoolValue(*projectSettings.IsCollectDatabaseSpecificsStatisticsEnabled),
		IsDataExplorerEnabled:                       types.BoolValue(*projectSettings.IsDataExplorerEnabled),
		IsExtendedStorageSizesEnabled:               types.BoolValue(*projectSettings.IsExtendedStorageSizesEnabled),
		IsPerformanceAdvisorEnabled:                 types.BoolValue(*projectSettings.IsPerformanceAdvisorEnabled),
		IsRealtimePerformancePanelEnabled:           types.BoolValue(*projectSettings.IsRealtimePerformancePanelEnabled),
		IsSchemaAdvisorEnabled:                      types.BoolValue(*projectSettings.IsSchemaAdvisorEnabled),
		Teams:                                       toTFTeamsDataSourceModel(ctx, teams),
		Limits:                                      toTFLimitsDataSourceModel(ctx, limits),
	}
```

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_projects.go:
```diff
@@ -0,0 +1,202 @@
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
```
```suggestion
	"go.mongodb.org/atlas-sdk/v20230201002/admin"
	matlas "go.mongodb.org/atlas/mongodbatlas"
	"github.com/hashicorp/terraform-plugin-framework/datasource"
	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
	"github.com/hashicorp/terraform-plugin-framework/types"
```

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_projects.go:
```diff
@@ -0,0 +1,202 @@
+
+type tfProjectsDSModel struct {
+	ID           types.String       `tfsdk:"id"`
+	Results      []tfProjectDSModel `tfsdk:"results"`
```
```suggestion
	Results      []*tfProjectDSModel `tfsdk:"results"`
```

maastha replied:
updated

andreaangiolillo commented on mongodbatlas/fw_data_source_mongodbatlas_projects_test.go:
```diff
@@ -6,22 +6,19 @@ import (
 	projectName := fmt.Sprintf("test-datasource-project-%s", acctest.RandString(10))
 	orgID := os.Getenv("MONGODB_ATLAS_ORG_ID")
 	teamsIds := strings.Split(os.Getenv("MONGODB_ATLAS_TEAMS_IDS"), ",")
-	apiKeysIds := strings.Split(os.Getenv("MONGODB_ATLAS_API_KEYS_IDS"), ",")
```
why did we remove this?

maastha replied:
Because we are also removing api_keys as part of this. Do you mean something else?

andreaangiolillo replied:
I forgot that we are deprecating apiKeysIds. Resolving this

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	"sort"
+	"time"
+
+	"go.mongodb.org/atlas-sdk/v20230201002/admin"
+	matlas "go.mongodb.org/atlas/mongodbatlas"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/booldefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/boolplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/setplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+
+	expanders "github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas/framework/expanders"
```
this seems to need a go imports

maastha replied:
I do have that setup. This is how it auto-formats all the imports for me. We ran into this last time too. Let's sync our goimports settings because this is auto-formatted :smile:

andreaangiolillo replied:
👍

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	client *MongoDBClient
+}
+
+type tfProjectRSModel struct {
```
```suggestion
type projectRSModel struct {
```

maastha replied:
Resolving as we will be using `tfResourceNameRSModel` as convention as discussed.

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	client *MongoDBClient
+}
+
+type tfProjectRSModel struct {
+	Limits                                      types.Set    `tfsdk:"limits"`
+	Teams                                       types.Set    `tfsdk:"teams"`
+	RegionUsageRestrictions                     types.String `tfsdk:"region_usage_restrictions"`
+	Name                                        types.String `tfsdk:"name"`
+	OrgID                                       types.String `tfsdk:"org_id"`
+	Created                                     types.String `tfsdk:"created"`
+	ProjectOwnerID                              types.String `tfsdk:"project_owner_id"`
+	ID                                          types.String `tfsdk:"id"`
+	ClusterCount                                types.Int64  `tfsdk:"cluster_count"`
+	IsDataExplorerEnabled                       types.Bool   `tfsdk:"is_data_explorer_enabled"`
+	IsPerformanceAdvisorEnabled                 types.Bool   `tfsdk:"is_performance_advisor_enabled"`
+	IsRealtimePerformancePanelEnabled           types.Bool   `tfsdk:"is_realtime_performance_panel_enabled"`
+	IsSchemaAdvisorEnabled                      types.Bool   `tfsdk:"is_schema_advisor_enabled"`
+	IsExtendedStorageSizesEnabled               types.Bool   `tfsdk:"is_extended_storage_sizes_enabled"`
+	IsCollectDatabaseSpecificsStatisticsEnabled types.Bool   `tfsdk:"is_collect_database_specifics_statistics_enabled"`
+	WithDefaultAlertsSettings                   types.Bool   `tfsdk:"with_default_alerts_settings"`
+}
```
Can't we do this? Or the opposite, the DS using the resource model
```suggestion
type tfProjectRSModel struct {
tfProjectDSModel
}
```

maastha replied:
No. Unfortunately there were differences in all models. For example, data source models don't have `WithDefaultAlertsSettings` or `ProjectOwnerID` fields and these models need to be exactly same as the schemas

andreaangiolillo replied:
Thanks!

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		return
+	}
+	r.client = client
+}
```
this seems used also by other ds and resources. Can we extract the implementation in another function and reuse it inside the Configure?

maastha replied:
Can change it like this, didn't seem to add a lot because of the error handling that's why didn't do it initially. But let me know.

```suggestion
func (r *ProjectRS) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
	client, err := configure(req.ProviderData)
	if err != nil {
		resp.Diagnostics.AddError(errorConfigureSummary, err.Error())
		return
	}
	r.client = client
}

func configure(providerData any) (*MongoDBClient, error) {
	if providerData == nil {
		return nil, fmt.Errorf("ProviderData is null")
	}
	client, ok := providerData.(*MongoDBClient)
	if !ok {
		return nil, fmt.Errorf(errorConfigure, providerData)
	}
	return client, nil
}
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		}
+		resp.Diagnostics.AddError(fmt.Sprintf("error updating project's settings assigned (%s):", project.ID), err.Error())
+		return
+	}
```
```suggestion
	if _, _, err = conn.Projects.UpdateProjectSettings(ctx, project.ID, projectSettings); err != nil {
		errd := deleteProject(ctx, r.client.Atlas, project.ID)
		if errd != nil {
			resp.Diagnostics.AddError("error during project deletion when updating project settings", fmt.Sprintf(errorProjectDelete, project.ID, err.Error()))
			return
		}
		resp.Diagnostics.AddError(fmt.Sprintf("error updating project's settings assigned (%s):", project.ID), err.Error())
		return
	}
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	}
+
+	// READ
+	// get project
```
```suggestion

```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	}
+
+	// READ
+	// get project
```
```suggestion
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+}
+
+func (r *ProjectRS) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
```
```suggestion
	resource.ImportStatePassThroughID(ctx, path.Root("id"), req, resp)
```

maastha replied:
This is an in-built terraform function

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+	}
+
+	// adding updated teams into the project
+	if len(planTeams) > 0 {
```
```suggestion
	if len(planTeams) == 0 {
	    return nil
	}
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		_, _, err := conn.Projects.AddTeamsToProject(ctx, projectID, toAtlasProjectTeams(ctx, planTeams))
+		if err != nil {
+			return fmt.Errorf("error adding teams to the project: %v", err.Error())
+		}
```
```suggestion
		if _, _, err := conn.Projects.AddTeamsToProject(ctx, projectID, toAtlasProjectTeams(ctx, planTeams)); err != nil {
			return fmt.Errorf("error adding teams to the project: %v", err.Error())
		}
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+
+If we get a defined error response, return that right away
+Else retry
+*/
```
```suggestion
resourceProjectDependentsDeletingRefreshFunc assumes the project CRUD outcome will be the same for any non-zero number of dependents

If all dependents are deleting, wait to try and delete
Else consider the aggregate dependents idle.

If we get a defined error response, return that right away
Else retry
*/
```
```suggestion
This assumes the project CRUD outcome will be the same for any non-zero number of dependents

If all dependents are deleting, wait to try and delete
Else consider the aggregate dependents idle.

If we get a defined error response, return that right away
Else retry
*/
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+			return nil, "", err
+		} else if err != nil {
+			return nil, "RETRY", nil
+		}
```
```suggestion
		if errors.As(err, &target) {
			return nil, "", err
		}
		
		if err != nil {
			return nil, "RETRY", nil
		}
```

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		clusters, _, err := client.AdvancedClusters.List(ctx, projectID, nil)
+		dependents := AtlastProjectDependents{AdvancedClusters: clusters}
+
+		if errors.As(err, &target) {
```
what is target? can we give it a better name?

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		if errors.As(err, &target) {
+			return nil, "", err
+		} else if err != nil {
+			return nil, "RETRY", nil
```
retry should be a constant. Also why are we returning it?

maastha replied:
This is called when deleting project. RETRY is returned so it can be checked as a status for a refresh function. See https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1365/files#diff-c2af32c41f07d77e630b76e890c8a1dfee1e053fdeac86f8fd31a0bb8e5c9663R801

This method migrated as is from SDKv2 resource

maastha replied:
updated to constant.

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		}
+
+		if dependents.AdvancedClusters.TotalCount == 0 {
+			return dependents, "IDLE", nil
```
IDLE should be a constant

maastha replied:
updated

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,857 @@
+		}
+
+		for _, v := range dependents.AdvancedClusters.Results {
+			if v.StateName != "DELETING" {
```
same

maastha replied:
updated

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project_migration_test.go:
```diff
@@ -0,0 +1,229 @@
+				),
+			},
+			{
+				ProtoV6ProviderFactories: testAccProviderV6Factories,
```
how this step will work? is the first step using the old provider and the second step the framework?

maastha replied:
Correct. First step will use version 1.11.0 of our provider that was published today.
Second step will create a plan using the Framework-based implementation of this resource/provider and check that no changes should be discovered as part of the plan.

andreaangiolillo commented on mongodbatlas/provider_test.go:
```diff
@@ -22,6 +24,9 @@ var testAccProviderV6Factories map[string]func() (tfprotov6.ProviderServer, erro
 // this provider instance has to be passed into mux server factory for its configure method to be invoked
 var testAccProviderSdkV2 *schema.Provider
 
+// Manually configured client client required for Framework-based acceptance tests
```
```suggestion
// testMongoDBClient is used to configure client required for Framework-based acceptance tests
```

AgustinBettati committed:
refactor: addressing PR comments for linting and code cleanup

AgustinBettati committed:
fix: define provider version in framework provider Metadata function

maastha committed:
Apply suggestions from code review

Co-authored-by: Andrea Angiolillo <andrea.angiolillo@mongodb.com>

AgustinBettati committed:
doc: add description for AWS Secrets Manager provider config attributes

AgustinBettati committed:
refacor mux provider fractory method name and comments

AgustinBettati commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,862 @@
+	errorProjectUpdate  = "error updating project (%s): %s"
+)
+
+var _ resource.Resource = &ProjectRS{}
+var _ resource.ResourceWithImportState = &ProjectRS{}
```
not sure if I see value in these statements, checking of correct interface implementation will be done as soon as resource is configured in the provider

maastha replied:
Could you please elaborate `as soon as resource is configured in the provider`?

AgustinBettati replied:
When the resource is added into the provider, it will have to comply with the resource.Resource interface: 
`return []func() resource.Resource{
		NewProjectRS,
}`


I do however see that for the ImportState interface there would be no compilation errors if something is incorrectly defined, so line 42 does have value.
To add, hashicorp example uses this convention: https://github.com/hashicorp/terraform-provider-scaffolding-framework/blob/main/internal/provider/example_resource.go#L22 so we can follow with it.

AgustinBettati committed:
doc: inlcude terraform version requirement in docs

AgustinBettati committed:
refactor: remove redundant else statement

maastha committed:
Address PR comments

maastha committed:
lint fix

maastha committed:
lint fix

andreaangiolillo reviewed (approved):
LGTM. Thanks for addressing all the comments. (Make sure that the failing tests is not related to your changes)

maastha committed:
improve tests

maastha committed:
minor

maastha committed:
minor

maastha committed:
Merge branch 'INTMDB-971' into INTMDB-908-migrate-project

maastha committed:
update data source tests

maastha commented:
> LGTM. Thanks for addressing all the comments. (Make sure that the failing tests is not related to your changes)

All project* acceptance tests passing.

Zuhairahmed commented on mongodbatlas/fw_data_source_mongodbatlas_project.go:
```diff
@@ -0,0 +1,255 @@
+package mongodbatlas
+
```
does "fw_" ahead of each of the file names stand for "framework"? if so suggest we call out somewhere in repo README file to let users know which resources / data sources have been migrated to new Terraform Plugin Framework

maastha replied:
@Zuhairahmed added to CONTRIBUTING.md

Zuhairahmed reviewed (approved):
minor suggestion comment, otherwise LGTM

AgustinBettati committed:
feat: new framework provider, main and acceptance tests to use mux server with existing sdk v2 provider (#1366)

* feat: define new framework provider, configure main and acceptance tests to use mux server with existing sdk v2 provider

* fix ConflictsWith attribute validation and parsing of list values in assume role schema

* refactor: addressing PR comments for linting and code cleanup

* fix: define provider version in framework provider Metadata function

* doc: add description for AWS Secrets Manager provider config attributes

* refactor mux provider fractory method name and comments

* doc: inlcude terraform version requirement in docs

* refactor: remove redundant else statement

* refactor: improve readability of methods in fw_provider + comments

maastha committed:
merge

maastha committed:
Merge branch 'CLOUDP-189585-plugin-framework-migration' into INTMDB-908-migrate-project

maastha committed:
Update CONTRIBUTING.md with Framework conventions

maastha committed:
Update CONTRIBUTING.md with Framework conventions

andreaangiolillo commented on mongodbatlas/fw_resource_mongodbatlas_project.go:
```diff
@@ -0,0 +1,941 @@
+	resp.Diagnostics.Append(resp.State.Set(ctx, &projectPlanNew)...)
+}
+
+func (r *ProjectRS) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
```
.
